include/SampleBuffer.h

```diff
@@ -269,6 +271,9 @@ class LMMS_EXPORT SampleBuffer : public QObject, public sharedObject
 	}
 
 
+	std::unique_ptr<OscillatorConstants::waveform_t> m_userAntiAliasWaveTable;
```

Comment:
Grepping through the sources, the `m_userAntiAliasWaveTable` does not come up in any plugins, so there do not seem to be any "external users".

My guess is that it was some sort of "semantic" decision perhaps? I.e. oscillator is used to generate samples, sample buffer is used to store them. So the final waveform is stored in a `SampleBuffer` class? I don't know much about `SampleBuffer` though, so I can't say if it's a good or a bad idea.

Reply:
I'll take a look at `SampleBuffer` to investigate this comment further.

Reply:
About my concerns:

1. Of course, the user defined wave must be outside `Oscillator::s_waveTables`, because you can have `N` user defined waves in LMMS, but only `1` triangle wave etc. And the idea to bundle them with `SampleBuffer` seems sane to me, since `SampleBuffer` is commonly used.
2. If the user defined wave does not have `OscillatorConstants::WAVETABLE_LENGTH`, this will not generate any out-of-bounds-read, since the only case of access is `sampleBuffer->userWaveSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);`, i.e. the values will be in range [0,1]. However, I think this violates the Nyquist-Shannon theorem? Imagine your user wave is very long (10 seconds). Then the few `WAVETABLE_LENGTH` (~2000) samples that you take here are rather random? FWIR, you would need to do a much much larger FFT/IFFT for such long waves (which could be annoying if you load songs with long samples?).
3. Having the wave not used in other plugins is fine, as long as the FFT/IFFT stay that small. However, me might increase them (point 2).

So only (2) is a real problem right now.

Reply:
On point 2, are you suggesting the oscillator should act like a repitching sampler, and with a 10-second sample at middle c, playback should consist of the 10-second sample on loop, other notes pitch adjusted?

Reply:
I'm not sure if what the design for 3OSC on very long "waves" is. Currently, if I load a 3 minutes wave into 3osc with master, playback seems broken, so I guess that is just not wanted (pure guess). I was mostly exaggerating, trying to illustrate the Nyquist-Shannon  problem. If your user wave already has double the size of the pre-defined waves, I guess you're violating the theorem.

So I'm currently trying to figure out what happens if the wave you load is larger than usual - does `SampleBuffer::update` truncate the wave to the correct size, or does it load the whole large wave? Or should we assume that users always create user waves of the correct size (I guess not)?

Reply:
My original intention was yes as per serum, if you load a wavetable it would need to be the correct size. This, of course, was a long time ago, However, serum offers a lot more than 3osc, and can allow for the processing of sampled audio into wavetables. Serum also has multiple waveforms in a file, but always of a set length, IIR 512, 1024 2048, with all waveforms within a single file the same length. I am not suggesting we have to follow serum, or should attempt to reproduce the wavetable producing and editing functionality.

If the user loads a single waveform of the correct length, then the Nyquist-Shannon limit is obeyed, and the waveform is played back as intended,  If the sample is longer than the expected length a  sample is formed by collecting every N samples, I would not know how to predict a sound for a given sample, but it would be reproducible with every load of that sample, and was intended as a way to give the user the option to use any audio file.  The sound produced, dependant on the source would probably contain many inharmonic overtones, and these are used when synthesizing 2d vibrating objects, such as drum skins. 

I hope these comments help give an insight into why I made the decisions I did and can be of use in any improvements.

A bit off topic, I am really impressed by all the improvements made to this PR, thank you for taking it on :)

Reply:
Interesting viewpoint with the usefulness of inharmonics.

I'd say the current behaviour is sane then. If anyone would later request that we also handle large waves "correctly" (i.e. being conform to Nyquist-Shannon), we could probably add a dropdown menu or a dialog which asks you if you want to use a large wavetable or not. This would mostly mean changing the `WAVETABLE_LENGTH` constants in  `Oscillator::generateAntiAliasUserWaveTable` into variables. Looks doable to me.

I guess we can resolve this comment then?

Reply:
I vote it is resolved. The longer wavetable implementation as you suggest should be handled in a future PR.

Reply:
Huh, interesting reading, thanks for sharing the details. I never actually realized (as a user) that a single-cycle waveform is the expected input for the "user-defined wave" button. I knew there are some pre-made waveform files that could be loaded, but I just assumed they are some sort of magic format I do not understand so I never examined it in detail.

I tried to generate 1.5 minutes long single cycle of a square wave in Audacity and loaded that as the user-defined wave. And it really does work even with such a big file (and turning wavetable oscillator mode on cleans it up nicely). When I previously loaded a 30 second long tone sweep, all I got was white noise in the standard mode (expected from the "random" sampling) and some sort of square-ish tone in the wavetable mode (not sure how it managed to extract _that_ from the file).

Maybe it is more of a user interface issue? I.e., perhaps the user should be told that a single-cycle wave is expected (or that anything that is loaded will be interpreted that way), and warned that a file may cause aliasing due to the simplified sampling when it is longer than _n_ samples. (Although _n_ may depend on the current sample rate, and may change during export?) So basically let's do it the Apple way and tell users "you are holding it wrong".. :))